jenkins/bootstrap.py

```diff
@@ -55,45 +56,91 @@
+        return False
+    if kill:  # Process will not die, kill everything
+        pgid = os.getpgid(proc.pid)
+        logging.info(
+            'Kill %d and process group %d', proc.pid, pgid)
+        os.killpg(pgid, signal.SIGKILL)
+        proc.kill()
+        return True
+    logging.info(
+        'Terminate %d on timeout', proc.pid)
+    proc.terminate()
+    return True
+
+
+def _call(end, cmd, stdin=None, check=True, output=None):
     """Start a subprocess."""
     logging.info('Call:  %s', ' '.join(pipes.quote(c) for c in cmd))
+    begin = time.time()
+    if end:
+        end = max(end, time.time() + 60)  # Allow at least 60s per command
+        logging.info('Limiting call to %.1f minutes', elapsed(begin))
     proc = subprocess.Popen(
         cmd,
         stdin=subprocess.PIPE if stdin is not None else None,
         stdout=subprocess.PIPE,
         stderr=subprocess.PIPE,
+        preexec_fn=os.setsid,
     )
     if stdin:
         proc.stdin.write(stdin)
         proc.stdin.close()
     out = []
     code = None
-    reads = [proc.stdout.fileno(), proc.stderr.fileno()]
+    timeout = False
+    reads = {
+        proc.stderr.fileno(): (proc.stderr, logging.warning),
+        proc.stdout.fileno(): (
+            proc.stdout, (out.append if output else logging.info)),
+    }
     while reads:
+        if terminate(end, proc, timeout):
+            if timeout:  # We killed everything
+                break
+            # Give subprocess some cleanup time before killing.
+            end = time.time() + 15 * 60
+            timeout = True
         ret = select.select(reads, [], [], 0.1)
         for fdesc in ret[0]:
-            if fdesc == proc.stdout.fileno():
-                line = proc.stdout.readline()
-                if not line:
-                    reads.remove(fdesc)
-                    continue
-                if output:
-                    out.append(line)
-                else:
-                    logging.info(line[:-1])
-            if fdesc == proc.stderr.fileno():
-                line = proc.stderr.readline()
-                if not line:
-                    reads.remove(fdesc)
-                    continue
-                logging.warning(line[:-1])
-        if not ret[0]:
-            if proc.poll() is not None:
-                break  # process exited without closing pipes (timeout?)
+            if read_all(end, *reads[fdesc]):
+                reads.pop(fdesc)
+        if not ret[0] and proc.poll() is not None:
+            break  # process exited without closing pipes (timeout?)
 
     code = proc.wait()
-    logging.info('process %d exited with code %d', proc.pid, code)
+    if timeout:
+        code = code or 124
+        logging.error('Build timed out')
+    if code:
+        logging.error('Build failed')
+    logging.info(
+        'process %d exited with code %d after %.1fm',
+        proc.pid, code, elapsed(begin))
+    out.append('')
     lines = output and '\n'.join(out)
```

Comment:
I think ignore `\n` is correct, thus it was wrong previously(`out.append(line)`)?